### PR TITLE
Dynamic Dashboard: Add banners to new dashboard screen

### LINKED ISSUE #12411: New enum for dashboard cards and update dashboard controller accordingly
Add a new enum for dashboard cards and create a SwiftUI version of the dashboard view controller to manage the UI more easily.

Checklist for dashboard view:

- [ ] Store name
- [ ] Tab bar item
- [ ] Share store
- [ ] Store stats and top performers
- [ ] Store onboarding
- [ ] Blaze
- [ ] Pull-to-refresh
- [ ] Top banner for error syncing
- [ ] JITM
- [ ] Bottom Jetpack benefit banner
- [ ] Announcements
- [ ] Web view sheet
- [ ] Store plan banner
- [ ] Privacy banner

### PR DESCRIPTION
**Pull request by itsmeichigo**
Closes: #12411 

## Description

This PR adds the missing banners from the refactoring in #12427:
- Offline banner
- Top banner for error syncing stats
- Bottom Jetpack benefit banner
- Store plan banner
- Privacy setting bottom sheet
- JITM modal
- JITM announcement banner
- JITM web view sheet.

## Testing instructions

**1. Offline Banner**
- Build the app to a physical device.
- Disable both wifi and cellular connections.
- Confirm that the offline banner is displayed on the dashboard screen.

**2. Top banner for error syncing stats**
- Repeat the steps in (1), confirm that an error banner is displayed at the top of the dashboard screen.
- Tap both CTAs on the banner, confirm that both works correctly.

**3. Bottom Jetpack benefit banner**
- Build the app and switch to a JCP site.
- Confirm that the Jetpack benefit banner is displayed at the bottom of the dashboard screen.
- Tap the banner, the Jetpack install flow should start.
- Dismiss the flow and tap the X button. Confirm that the banner is dismissed.

**4. Store plan banner**
- Build the app and switch to a free trial site (or create a new one).
- Confirm that the banner for the site plan status is displayed at the bottom of the screen.

**5. Privacy setting bottom sheet**
- In `PrivacyBannerPresenter` comment out the following `if` check:

https://github.com/woocommerce/woocommerce-ios/blob/2244abb16a73bb0f307b44899c482a5115c89f2b/WooCommerce/Classes/ViewRelated/Dashboard/Settings/Privacy/PrivacyBannerPresenter.swift#L33-L35

- Build the app and confirm that the privacy bottom sheet is presented correctly.

**6. JITM modal**
- Follow the testing steps in #9694, please remember to replace `{EPHEMERAL-SITE-HOST}` in the sample response with your site's host.
- Confirm that the JITM modal is displayed correctly.

**7. JITM announcement banner**
- Follow the testing steps in #9734.
- Confirm that the JITM feature announcement banner is displayed correctly.

**8. JITM web view sheet**
- Follow the testing steps in pdfdoF-1uc-p2#comment-2581.
- Tap on Purchase a card reader on the JITM feature announcement banner.
- Confirm that a web view for purchasing a card reader is displayed.

https://github.com/woocommerce/woocommerce-ios/assets/5533851/1ceb2a9b-1a6c-414f-9340-8c8ace7aabd6

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/Classes/ViewRelated/Top Banner/ErrorTopBanner.swift [hunk 1]
```diff
@@ -0,0 +1,83 @@
+import SwiftUI
+
+/// SwiftUI view equivalent to the view created by `ErrorTopBannerFactory`.
+/// TODO: if needed - convert this view to a reusable top banner.
+struct ErrorTopBanner: View {
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Top Banner/ErrorTopBanner.swift:5]: I tried wrapping `ErrorTopBannerFactory`'s view in a `UIViewRepresentable` but the view could not resolve the constraints properly when added to the dashboard view. I decided to rewrite the banner in SwiftUI instead as it's a fairly simple view.

File: WooCommerce/Classes/ViewRelated/Dashboard/DashboardView.swift [hunk 2]
```diff
@@ -7,6 +7,11 @@ import struct Yosemite.StoreOnboardingTask
 struct DashboardView: View {
     @ObservedObject private var viewModel: DashboardViewModel
     @State private var currentSite: Site?
+    @State private var dismissedJetpackBenefitBanner = false
+    @State private var showingSupportForm = false
+    @State private var troubleShootURL: URL?
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Dashboard/DashboardView.swift]: Super nit
```suggestion
    @State private var troubleshootURL: URL?
```

File: WooCommerce/Classes/ViewRelated/Dashboard/DashboardView.swift [hunk 3]
```diff
@@ -58,13 +92,36 @@ struct DashboardView: View {
...
                 await viewModel.reloadAllData()
                 await storeStatsAndTopPerformersViewController.reloadData(forced: true)
             }
         }
+        .safeAreaInset(edge: .bottom) {
+            jetpackBenefitBanner
+                .renderedIf(shouldShowJetpackBenefitsBanner)
+
+            storePlanBanner
+                .renderedIf(connectivityStatus != .notReachable)
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Dashboard/DashboardView.swift:113]: nit: Can we move this `connectivityStatus != .notReachable` check into `storePlanBanner` property as part of the `if` condition? https://github.com/woocommerce/woocommerce-ios/pull/12433/files#diff-9b48c8193ddc5598342e536bc43291d4e4885152fbf46aac277d4c6de29b27b1R195.

File: WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewHostingController.swift [hunk 4]
```diff
@@ -44,6 +61,33 @@ private extension DashboardViewHostingController {
...
         tabBarItem.accessibilityIdentifier = "tab-bar-my-store-item"
     }
+
+    /// Presents the privacy banner if needed.
+    ///
+    func presentPrivacyBannerIfNeeded() {
+        privacyBannerPresenter.presentIfNeeded(from: self)
+    }
+
+    private func observeModalJustInTimeMessages() {
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewHostingController.swift]: nit: We can remove the `private` keyword as this method is inside a private extension.

File: WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift [hunk 5]
```diff
@@ -28,12 +28,15 @@ final class DashboardViewModel: ObservableObject {
 
     let blazeCampaignDashboardViewModel: BlazeCampaignDashboardViewModel
 
-    @Published private(set) var showWebViewSheet: WebViewSheetViewModel? = nil
+    @Published var showWebViewSheet: WebViewSheetViewModel? = nil
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift]: nit: `showWebViewSheet` sounds like a Bool, but this property holds a view model. Should we consider renaming this?

File: WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift [hunk 6]
```diff
@@ -250,6 +256,21 @@ final class DashboardViewModel: ObservableObject {
...
     }
+
+    func saveJetpackBenefitBannerDismissedTime() {
+        let dismissAction = AppSettingsAction.setJetpackBenefitsBannerLastDismissedTime(time: Date())
+        stores.dispatch(dismissAction)
+    }
+
+    func maybeSyncAnnouncementsAfterWebViewDismissed() {
+        // If the web view was opened from a modal JITM, it was dismissed before the webview
+        // was presented. Syncing in that situation would result in it showing again.
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift]: If I understand this correctly,
```suggestion
        // was presented. Syncing in that situation would result in the JITM model showing again.
```

**Review by selanthiraiyan — APPROVED** [review]
Thanks for the detailed testing instructions, Huong! All the testing steps work as expected. 🎉 

The code looks good as well. I left a few non-blocking questions/nits. 🚢

File: WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift (hunk 5 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift]: Good point - updated in 9b6007d20be052455815c2faf4cf78c46b09f308.

File: WooCommerce/Classes/ViewRelated/Dashboard/DashboardView.swift (hunk 2 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Dashboard/DashboardView.swift]: Updated in b039930e86b08a6e17add6a8054b34711ac7d137.

File: WooCommerce/Classes/ViewRelated/Dashboard/DashboardView.swift (hunk 3 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Dashboard/DashboardView.swift:113]: I intentionally keep the  check this way to make it clear when we don't want to display the store plan banner. I hope that makes sense.

File: WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewHostingController.swift (hunk 4 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewHostingController.swift]: Removed in df352f1eef0fc7c59c21a1b52b353f51c3945293.

File: WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift (hunk 6 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift]: I copied the whole method as-is from the old `DashboardViewController`. Updated the wording in 8141bef2f6e9a0591cb051b784bdd8ea47561aa6, I hope that's correct.

**Comment by itsmeichigo** [comment]
Thank you @selanthiraiyan for the reviews! I responded to your comments - enabling auto-merge now to avoid blocking development.

**Pull request merged**